pricegraph/src/orderbook.rs

```diff
@@ -156,45 +166,37 @@ impl Orderbook {
-            bellman_ford::search(&self.projection, quote)
-        {
-            let path = path::find_cycle(&predecessors, node, Some(quote))
-                .expect("negative cycle not found after being detected");
-
-            let base_index_search = if path.first() == Some(&quote) {
-                path.iter().position(|node| *node == base)
-            } else {
-                // NOTE: If the path doesn't start with the quote token, don't
-                // even need to search for the base token.
-                None
-            };
-            let base_index = match base_index_search {
-                Some(value) => value,
-                None => {
-                    // NOTE: Skip negative cycles that are not along the
-                    // specified market.
-                    self.fill_path(&path).unwrap_or_else(|| {
-                        panic!(
-                            "failed to fill path along detected negative cycle {}",
-                            format_path(&path),
-                        )
-                    });
-
-                    continue;
+        while let Err(cycle) = bellman_ford::search(&self.projection, quote) {
+            let cycle_not_in_path = match cycle.with_starting_node(quote) {
+                Err(cycle) => cycle,
+                Ok(cycle) => {
+                    let base_index_search = cycle.0.iter().position(|node| *node == base);
+                    match base_index_search {
+                        None => cycle,
+                        Some(base_index) => {
+                            let (ask, bid) = (&cycle.0[0..base_index + 1], &cycle.0[base_index..]);
+
+                            let ask = self
+                                .fill_path(ask)
+                                .expect("ask transitive path not found after being detected");
+                            let bid = self
+                                .fill_path(bid)
+                                .expect("bid transitive path not found after being detected");
+
+                            return Some(Ring { ask, bid });
+                        }
+                    }
                 }
             };
 
-            let (ask, bid) = (&path[0..base_index + 1], &path[base_index..]);
-
-            let ask = self
-                .fill_path(ask)
-                .expect("ask transitive path not found after being detected");
-            let bid = self
-                .fill_path(bid)
-                .expect("bid transitive path not found after being detected");
-
-            return Some(Ring { ask, bid });
+            // NOTE: Skip negative cycles that are not along the
+            // specified market.
+            self.fill_path(&cycle_not_in_path.0).unwrap_or_else(|| {
+                panic!(
+                    "failed to fill path along detected negative cycle {}",
+                    format_path(&cycle_not_in_path.0),
+                )
+            });
```

Comment:
Can we move this into the Err() arm of the match? It took me a while to understand how we would get here and I think it would clearer to treat it right next to the condition.

It could be something like:

```rust
match cycle.with_starting_node(quote).map(|cycle| cycle.0.iter().position(|node| *node == base)) {
  Err(cycle_not_in_path) => ...
  Some(base_index) => ...
}
```

Reply:
The suggestion doesn't work as it is because `cycle` is consumed in the process and i need it inside the `Some` branch, but I'll think of something.
If you think it's more readable I can probably make it look more like it was before my changes, I remember I was quite confused by how it was written before, but maybe my changes didn't improve the situation.

Reply:
Would this become possible if cycle implemented `split` directly and would give you an `Option<(before, after)>` instead of `Option<base_index>`? Then you shouldn't need to access the original cycle in the some block

Reply:
I did something similar, you can check if you find the result more readable.